### PR TITLE
Create rudementary powerplant simulation

### PR DESCRIPTION
**Pull request by gidoBOSSftw5731**
See #27

### CODE REVIEW

File: powerplant/main.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+// updateOutput updates the powerplant output based on the target and reactivity.
+// The reactivity is added to the sine of the current time (minutes in the hour,
+// scaled to 0-1) to get the new output.
+func (p *powerplant) updateOutput() {
+	for {
+		p.output = p.output + p.reactivity + float32(.2*
+			math.Sin((float64(
+				time.Now().Minute())+(float64(time.Now().Second())/
+				60.0))/60.0*2*math.Pi))
+		log.Println("current output:", p.output)
```
Comment (morrowc) [inline powerplant/main.go:138]: you may not want this to log 'all the time'.

File: controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	"math"
+	"time"
+
+	"github.com/simonvetter/modbus"
+)
+
+const (
+	targetpower = 100
+	// margin is 10% of safety range
+	//margin = float32(20 * .1)
```
Comment (morrowc) [inline controller/controller.go:14]: random left over comment?

File: controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	margin               = 0
+	adjustment_step_size = 5
+)
+
+func main() {
+	// init modbus client
+	client, err := modbus.NewClient(&modbus.ClientConfiguration{
+		URL: "tcp://localhost:5502",
```
Comment (morrowc) [inline controller/controller.go:22]: should make this a var or a flag with a default value.

File: powerplant/main.go (hunk 1 above)
Comment (gidoBOSSftw5731) [reply powerplant/main.go:138]: For now it makes sense to leave it for debugging, all of this will just be shoved in to graylog anyway, at which point it's filterable.

File: controller/controller.go (hunk 2 above)
Comment (gidoBOSSftw5731) [reply controller/controller.go:14]: Margin ended up not being used for now, but may be reimplemented later if more safety is desired.

File: powerplant/main.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	// Operating range is always 10% of safety.
+	safety float32
+}
+
+var (
+	// plant is the powerplant object
+	// todo: store this in modbus somewhere
+	plant = &powerplant{
+		reactivity: 0,
+		output:     100,
```
Comment (ashim-mahara) [inline powerplant/main.go:32]: We have to introduce an equation to calculate the output of the powerplant based on reactivity (or control rod position if we add them to the equation). It looks awkward to output 100 when no reactivity or any other parameters are controlling it.

File: powerplant/main.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	handler := &modbusHandler{}
+
+	// create the server object
+	server, err = modbus.NewServer(&modbus.ServerConfiguration{
+		// listen on localhost port 5502
+		URL: "tcp://0.0.0.0:5502",
+		// close idle connections after 30s of inactivity
+		Timeout: 30 * time.Second,
+		// accept 5 concurrent connections max.
+		MaxClients: 5,
```
Comment (ashim-mahara) [inline powerplant/main.go:52]: We can further limit the number of clients to introduce some limits and intended paths.

File: powerplant/main.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	return nil, nil
+}
+
+// Holding registers are used for both input and output of data.
+func (h *modbusHandler) HandleHoldingRegisters(req *modbus.HoldingRegistersRequest) (
+	res []uint16, err error) {
+	log.Printf("Holding Registers request: %#v", *req)
+	// This switch calls go functions for any modbus function desired.
+	switch req.Addr {
+	case 65:
```
Comment (ashim-mahara) [inline powerplant/main.go:83]: Are the choices 65 and 67 arbitrary?

File: powerplant/main.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+		time.Sleep(1 * time.Second)
+	}
+}
+
+//
+
+// generateOutput does the math which is used by updateOutput
+func (p *powerplant) generateOutput(t time.Time) float32 {
+	return p.output + p.reactivity + float32(.2*
+		math.Sin((float64(
```
Comment (ashim-mahara) [inline powerplant/main.go:146]: Why multiply by the equation?

**Review by ashim-mahara — COMMENTED** [review]
You can address the changes in this PR or another PR.

File: controller/controller.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+		time.Sleep(1 * time.Second)
+
+		// get the data from the powerplant and parse it out
+		out, err := client.ReadRegisters(65, 2, modbus.HOLDING_REGISTER)
+		if err != nil {
+			log.Println("Error reading register", err)
+			continue
+		}
+		// convert to float32
+		outbits := uint32(out[0])<<16 | uint32(out[1])
```
[the remark this answers is not in the document]
Comment (gidoBOSSftw5731) [reply controller/controller.go:45]: Sure, if it's run as a service it'll automatically be dumped into graylog. I'll have to look into what's the best way to format the output.

File: powerplant/main.go (hunk 4 above)
Comment (gidoBOSSftw5731) [reply powerplant/main.go:32]: Hmm, maybe have some sort of system where there's energy being created vs energy being used, will investigate further.

File: powerplant/main.go (hunk 5 above)
Comment (gidoBOSSftw5731) [reply powerplant/main.go:52]: Yep, this was just the default, it's all tunable.

File: powerplant/main.go (hunk 6 above)
Comment (gidoBOSSftw5731) [reply powerplant/main.go:83]: No, they're standard numbers. 66 is implied with 65 because you always request 2 16bit registers.

File: powerplant/main.go (hunk 7 above)
Comment (gidoBOSSftw5731) [reply powerplant/main.go:146]: I was playing with the numbers to get a result that was manageable and not too variable, I settled on .2 as it'll show some fluctuations while staying under control while also going outside of the safety margins fairly quickly if not kept in check.

File: controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	"flag"
+	"log"
+	"math"
+	"time"
+
+	"github.com/simonvetter/modbus"
+)
+
+const (
+	targetpower = 100
```
[the remark this answers is not in the document]
Comment (gidoBOSSftw5731) [reply controller/controller.go:13]: This actually isn't that realistic. Nuclear plants are base loads that run at 99% capacity 24/7, while other "peaker" plants vary their output throughout the day to meet different demands (think, city water pumps always need power, etc)

**Comment by ashim-mahara** [comment]
LGTM.

Further tasks:

1. Provision system, install service, write to graylog.
2. Introduce control rods or reactivity to calculate output.
3. Variable moving target.

File: controller/controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ashim-mahara) [reply controller/controller.go:13]: https://www.reddit.com/r/NuclearPower/comments/4xbxu6/to_what_extent_are_nuclear_power_plants/

This looks like a good discussion on the subject. 

TLDR; NPP may load follow to a certain extent, maybe 10-20%, but they're designed to be run at 100% as you said. We could put an arbitrary number as the target power and let the controller try to reach that target with the randomness introduced with the 0.2 * math.sine equation.

**Comment by ashim-mahara** [comment]
@gidoBOSSftw5731 I am just going to merge this since all of the code appears to be working. Feel free to address the above in another PR.
